File: online2/metrics_logger.py
"""
Structured CSV logging for Online2 solver executions.
"""

import csv
import os
import threading
import time
from typing import Dict, List, Optional


class SolverMetricsLogger:
    """Thread-safe CSV logger for solver run metrics."""

    RUN_FIELDS = [
        "run_id",
        "run_sequence",
        "current_slot",
        "pending_batch_size",
        "total_assignments",
        "new_assignments",
        "replanned_assignments",
        "solver_status",
        "solver_mode",
        "lock_future_assignments",
        "solver_start_ts",
        "solver_end_ts",
        "solver_elapsed_ms",
        "avg_ms_per_new_request",
        "avg_ms_per_assignment",
        "total_carbon_cost",
        "carbon_cost_per_new_request",
        "carbon_cost_per_assignment",
        "error_window_avg_after",
        "error_window_avg_after_real",
        "error_window_start_slot",
        "error_window_end_slot",
        "error_window_threshold",
        "error_window_violated_after",
        "error_window_violated_after_real",
        "batches_processed_after",
        "total_scheduled_after",
    ]

    ASSIGNMENT_FIELDS = [
        "run_id",
        "current_slot",
        "solver_start_ts",
        "solver_end_ts",
        "request_id",
        "is_pending_request",
        "is_new_assignment_in_run",
        "scheduled_slot",
        "strategy_name",
        "strategy_duration",
        "error",
        "carbon_cost",
        "arrival_slot",
        "deadline_slot",
    ]

    SLOT_METRIC_FIELDS = [
        "run_id",
        "current_slot",
        "scheduled_slot",
        "run_slot_count",
        "total_slot_count_after",
        "avg_error_in_slot",
        "run_avg_error_in_slot",
        "slot_has_assignments_after",
        "carbon_intensity",
        "capacity_multiplier_after",
        "capacity_level_max_requests",
        "request_ids",
        "strategy_breakdown",
    ]

    INFEASIBLE_DEBUG_FIELDS = [
        "event_id",
        "timestamp",
        "current_slot",
        "pending_batch_size",
        "pending_request_details",
        "strict_threshold",
        "baseline_error_sum",
        "baseline_request_count",
        "baseline_average_error",
        "min_strategy_error",
        "max_strategy_error",
        "min_possible_avg_error_pending_only",
        "max_possible_avg_error_pending_only",
        "strict_infeasible_by_error_bound",
        "strict_scheduled_pending_count",
        "relaxed_scheduled_pending_count",
        "relaxed_success",
        "lock_future_assignments",
        "future_assignment_count",
        "future_slot_counts",
        "future_assignment_details",
        "all_active_slot_counts",
    ]

    def __init__(
        self,
        enabled: bool,
        runs_file: str,
        assignments_file: str,
        slot_metrics_file: str,
        infeasible_debug_file: Optional[str] = None,
    ) -> None:
        self.enabled = enabled
        self.runs_file = runs_file
        self.assignments_file = assignments_file
        self.slot_metrics_file = slot_metrics_file
        self.infeasible_debug_file = infeasible_debug_file
        self._lock = threading.Lock()

        if self.enabled:
            self._ensure_file_with_header(self.runs_file, self.RUN_FIELDS)
            self._ensure_file_with_header(self.assignments_file, self.ASSIGNMENT_FIELDS)
            self._ensure_file_with_header(self.slot_metrics_file, self.SLOT_METRIC_FIELDS)
            if self.infeasible_debug_file:
                self._ensure_file_with_header(self.infeasible_debug_file, self.INFEASIBLE_DEBUG_FIELDS)

    def log_solver_run(
        self,
        run_data: Dict,
        assignment_rows: Optional[List[Dict]] = None,
        slot_metric_rows: Optional[List[Dict]] = None,
    ) -> str:
        """
        Persist one solver run and its detailed rows.

        Returns:
            run_id used for this execution.
        """
        if not self.enabled:
            return ""

        assignment_rows = assignment_rows or []
        slot_metric_rows = slot_metric_rows or []

        run_id = run_data.get("run_id")
        if not run_id:
            run_id = f"{int(time.time() * 1000)}-{run_data.get('current_slot', 0)}"
            run_data["run_id"] = run_id

        with self._lock:
            self._append_row(self.runs_file, self.RUN_FIELDS, run_data)
            if assignment_rows:
                for row in assignment_rows:
                    row["run_id"] = run_id
                self._append_rows(self.assignments_file, self.ASSIGNMENT_FIELDS, assignment_rows)
            if slot_metric_rows:
                for row in slot_metric_rows:
                    row["run_id"] = run_id
                self._append_rows(self.slot_metrics_file, self.SLOT_METRIC_FIELDS, slot_metric_rows)
        return run_id

    def log_infeasible_debug(self, event_data: Dict) -> str:
        """
        Persist one strict-infeasibility debug event.

        Returns:
            event_id used for this debug row.
        """
        if not self.enabled or not self.infeasible_debug_file:
            return ""

        event_id = event_data.get("event_id")
        if not event_id:
            event_id = f"{int(time.time() * 1000)}-{event_data.get('current_slot', 0)}"
            event_data["event_id"] = event_id
        if "timestamp" not in event_data:
            event_data["timestamp"] = time.time()

        with self._lock:
            self._append_row(self.infeasible_debug_file, self.INFEASIBLE_DEBUG_FIELDS, event_data)
        return event_id

    def _ensure_file_with_header(self, path: str, header: List[str]) -> None:
        directory = os.path.dirname(path)
        if directory:
            os.makedirs(directory, exist_ok=True)
        must_create = not os.path.exists(path) or os.path.getsize(path) == 0
        if not must_create:
            with open(path, "r", newline="") as f:
                reader = csv.reader(f)
                existing_header = next(reader, [])
            must_create = existing_header != header
            if must_create:
                backup_path = f"{path}.bak.{int(time.time())}"
                os.replace(path, backup_path)

        if must_create:
            with open(path, "w", newline="") as f:
                writer = csv.DictWriter(f, fieldnames=header)
                writer.writeheader()

    def _append_row(self, path: str, header: List[str], row: Dict) -> None:
        self._append_rows(path, header, [row])

    def _append_rows(self, path: str, header: List[str], rows: List[Dict]) -> None:
        with open(path, "a", newline="") as f:
            writer = csv.DictWriter(f, fieldnames=header)
            for row in rows:
                normalized = {k: row.get(k, "") for k in header}
                writer.writerow(normalized)

File: rust/src/metrics_logger.rs
/// Thread-safe CSV metrics logger.
///
/// Mirrors `metrics_logger.py::SolverMetricsLogger`.
///
/// All files are opened for append on each write (no persistent file handles)
/// and serialised through a single `Mutex`.  This keeps the API Send+Sync
/// without lifetime complexity.

use std::collections::HashMap;
use std::fs::OpenOptions;
use std::io::BufRead;
use std::path::Path;
use std::sync::Mutex;
use std::time::{SystemTime, UNIX_EPOCH};

// ─── field order ─────────────────────────────────────────────────────────────

const RUN_FIELDS: &[&str] = &[
    "run_id",
    "run_sequence",
    "current_slot",
    "pending_batch_size",
    "total_assignments",
    "new_assignments",
    "replanned_assignments",
    "solver_status",
    "solver_mode",
    "lock_future_assignments",
    "solver_start_ts",
    "solver_end_ts",
    "solver_elapsed_ms",
    "avg_ms_per_new_request",
    "avg_ms_per_assignment",
    "total_carbon_cost",
    "carbon_cost_per_new_request",
    "carbon_cost_per_assignment",
    "error_window_avg_after",
    "error_window_avg_after_real",
    "error_window_start_slot",
    "error_window_end_slot",
    "error_window_threshold",
    "error_window_violated_after",
    "error_window_violated_after_real",
    "batches_processed_after",
    "total_scheduled_after",
    "global_error_before",
    "global_error_count_before",
    "global_error_constraint_active",
];

const ASSIGNMENT_FIELDS: &[&str] = &[
    "run_id",
    "current_slot",
    "solver_start_ts",
    "solver_end_ts",
    "request_id",
    "is_pending_request",
    "is_new_assignment_in_run",
    "scheduled_slot",
    "flavour_name",
    "flavour_duration",
    "error",
    "carbon_cost",
    "arrival_slot",
    "deadline_slot",
];

const SLOT_METRIC_FIELDS: &[&str] = &[
    "run_id",
    "current_slot",
    "scheduled_slot",
    "run_slot_count",
    "total_slot_count_after",
    "avg_error_in_slot",
    "run_avg_error_in_slot",
    "slot_has_assignments_after",
    "carbon_intensity",
    "capacity_multiplier_after",
    "capacity_level_max_requests",
    "request_ids",
    "flavour_breakdown",
];

const INFEASIBLE_DEBUG_FIELDS: &[&str] = &[
    "event_id",
    "timestamp",
    "current_slot",
    "pending_batch_size",
    "pending_request_details",
    "strict_threshold",
    "baseline_error_sum",
    "baseline_request_count",
    "baseline_average_error",
    "min_flavour_error",
    "max_flavour_error",
    "min_possible_avg_error_pending_only",
    "max_possible_avg_error_pending_only",
    "strict_infeasible_by_error_bound",
    "strict_scheduled_pending_count",
    "relaxed_scheduled_pending_count",
    "relaxed_success",
    "lock_future_assignments",
    "future_assignment_count",
    "future_slot_counts",
    "future_assignment_details",
    "all_active_slot_counts",
];

// ─── MetricsLogger ───────────────────────────────────────────────────────────

pub struct MetricsLogger {
    pub enabled: bool,
    pub runs_file: String,
    pub assignments_file: String,
    pub slot_metrics_file: String,
    pub infeasible_debug_file: Option<String>,
    lock: Mutex<()>,
}

impl MetricsLogger {
    pub fn new(
        enabled: bool,
        runs_file: String,
        assignments_file: String,
        slot_metrics_file: String,
        infeasible_debug_file: Option<String>,
    ) -> Self {
        let logger = Self {
            enabled,
            runs_file,
            assignments_file,
            slot_metrics_file,
            infeasible_debug_file,
            lock: Mutex::new(()),
        };
        if logger.enabled {
            ensure_header(&logger.runs_file, RUN_FIELDS);
            ensure_header(&logger.assignments_file, ASSIGNMENT_FIELDS);
            ensure_header(&logger.slot_metrics_file, SLOT_METRIC_FIELDS);
            if let Some(ref p) = logger.infeasible_debug_file {
                ensure_header(p, INFEASIBLE_DEBUG_FIELDS);
            }
        }
        logger
    }

    /// Log one solver run plus its per-assignment and per-slot-metric rows.
    ///
    /// Returns the generated `run_id` string (empty string when disabled).
    pub fn log_solver_run(
        &self,
        run_data: &HashMap<String, String>,
        assignment_rows: &[HashMap<String, String>],
        slot_metric_rows: &[HashMap<String, String>],
    ) -> String {
        if !self.enabled {
            return String::new();
        }
        let run_id = run_id_from_data(run_data);
        let _guard = self.lock.lock().unwrap();
        let mut full_run = run_data.clone();
        full_run.insert("run_id".to_string(), run_id.clone());
        append_rows(&self.runs_file, RUN_FIELDS, &[full_run]);

        let augmented_assignments: Vec<HashMap<String, String>> = assignment_rows
            .iter()
            .map(|r| {
                let mut row = r.clone();
                row.insert("run_id".to_string(), run_id.clone());
                row
            })
            .collect();
        append_rows(&self.assignments_file, ASSIGNMENT_FIELDS, &augmented_assignments);

        let augmented_slots: Vec<HashMap<String, String>> = slot_metric_rows
            .iter()
            .map(|r| {
                let mut row = r.clone();
                row.insert("run_id".to_string(), run_id.clone());
                row
            })
            .collect();
        append_rows(&self.slot_metrics_file, SLOT_METRIC_FIELDS, &augmented_slots);
        run_id
    }

    /// Log one infeasibility debug event. Returns the event_id string.
    pub fn log_infeasible_debug(&self, event_data: &HashMap<String, String>) -> String {
        if !self.enabled {
            return String::new();
        }
        if let Some(ref path) = self.infeasible_debug_file {
            let event_id = format!("inf_{}", unix_now_ms());
            let _guard = self.lock.lock().unwrap();
            let mut row = event_data.clone();
            row.insert("event_id".to_string(), event_id.clone());
            row.insert("timestamp".to_string(), unix_now_ms().to_string());
            append_rows(path, INFEASIBLE_DEBUG_FIELDS, &[row]);
            return event_id;
        }
        String::new()
    }
}

// ─── helpers ─────────────────────────────────────────────────────────────────

fn unix_now_ms() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis() as u64
}

/// Generate a run_id from the run_data map (slot + sequence).
fn run_id_from_data(data: &HashMap<String, String>) -> String {
    let slot = data.get("current_slot").map(|s| s.as_str()).unwrap_or("0");
    let seq = data.get("run_sequence").map(|s| s.as_str()).unwrap_or("0");
    let ts = unix_now_ms();
    format!("run_s{slot}_n{seq}_{ts}")
}

/// Ensure `path` starts with the expected CSV header row.
///
/// If the file is missing or has a different header, it is backed up and a
/// fresh file is created with the correct header.
fn ensure_header(path: &str, fields: &[&str]) {
    if let Some(parent) = Path::new(path).parent() {
        if !parent.as_os_str().is_empty() {
            std::fs::create_dir_all(parent).ok();
        }
    }

    let expected = fields.join(",");
    let needs_create = if Path::new(path).exists() {
        let file = std::fs::File::open(path).ok();
        let first_line = file.and_then(|f| {
            let mut reader = std::io::BufReader::new(f);
            let mut line = String::new();
            reader.read_line(&mut line).ok()?;
            Some(line.trim().to_string())
        });
        first_line.map_or(true, |h| h != expected)
    } else {
        true
    };

    if needs_create {
        if Path::new(path).exists() {
            let backup = format!("{}.bak.{}", path, unix_now_ms());
            std::fs::rename(path, &backup).ok();
        }
        let mut wtr = csv::Writer::from_path(path).expect("Cannot create CSV file");
        wtr.write_record(fields).ok();
        wtr.flush().ok();
    }
}

/// Append `rows` to an existing CSV file (no header written).
fn append_rows(path: &str, fields: &[&str], rows: &[HashMap<String, String>]) {
    if rows.is_empty() {
        return;
    }
    let file = OpenOptions::new()
        .append(true)
        .create(true)
        .open(path)
        .expect("Cannot open CSV for append");

    let mut wtr = csv::WriterBuilder::new()
        .has_headers(false)
        .from_writer(file);

    for row in rows {
        let record: Vec<&str> = fields
            .iter()
            .map(|f| row.get(*f).map(|s| s.as_str()).unwrap_or(""))
            .collect();
        wtr.write_record(&record).ok();
    }
    wtr.flush().ok();
}
